pyradiosky/skymodel.py

```diff
@@ -809,55 +1261,212 @@ def read_votable_catalog(gleam_votable, source_select_kwds={}, return_table=Fals
+
+    # Convert to astropy Table
+    astropy_table = table_match.to_table()
+
+    # get ID column
+    id_col_use = _get_matching_fields(id_column, astropy_table.colnames)
+
+    # get RA & Dec columns, if multiple matches, exclude VizieR calculated columns
+    # which start with an underscore
+    ra_col_use = _get_matching_fields(
+        ra_column, astropy_table.colnames, exclude_start_pattern="_"
+    )
+    dec_col_use = _get_matching_fields(
+        dec_column, astropy_table.colnames, exclude_start_pattern="_"
+    )
+
+    if isinstance(flux_columns, (str)):
+        flux_columns = [flux_columns]
+    flux_cols_use = []
+    for col in flux_columns:
+        flux_cols_use.append(_get_matching_fields(col, astropy_table.colnames))
+
+    if len(flux_columns) > 1 and freq_array is None:
+        raise ValueError("freq_array must be provided for multiple flux columns.")
+
+    if reference_frequency is not None or len(flux_cols_use) == 1:
+        if reference_frequency is not None:
+            if not isinstance(reference_frequency, (Quantity,)):
+                raise ValueError("reference_frequency must be an astropy Quantity.")
+            reference_frequency = (
+                np.array([reference_frequency.value] * len(astropy_table))
+                * reference_frequency.unit
+            )
+        if spectral_index_column is not None:
+            spectral_type = "spectral_index"
+            spec_index_col_use = _get_matching_fields(
+                spectral_index_column, astropy_table.colnames
+            )
+            spectral_index = astropy_table[spec_index_col_use].data.data
+        else:
+            spectral_type = "flat"
+            spectral_index = None
+    else:
+        spectral_type = "subband"
+        spectral_index = None
+
+    stokes = np.zeros((4, len(flux_cols_use), len(astropy_table)), dtype=np.float)
+    for index, col in enumerate(flux_cols_use):
+        stokes[0, index, :] = astropy_table[col].data.data
+
+    skymodel_obj = SkyModel(
+        astropy_table[id_col_use].data.data.astype("str"),
+        Longitude(astropy_table[ra_col_use].data.data, astropy_table[ra_col_use].unit),
+        Latitude(astropy_table[dec_col_use].data.data, astropy_table[dec_col_use].unit),
+        stokes,
+        spectral_type,
+        freq_array=freq_array,
+        reference_frequency=reference_frequency,
+        spectral_index=spectral_index,
+    )
+
+    if source_select_kwds is not None:
+        sky_array = skymodel_to_array(skymodel_obj)
+        sky_array = source_cuts(sky_array, **source_select_kwds)
+        if return_table:
+            return sky_array
+
+        skymodel_obj = array_to_skymodel(sky_array)
 
     if return_table:
-        return data
+        return skymodel_to_array(skymodel_obj)
 
-    return array_to_skymodel(data)
+    return skymodel_obj
 
 
-def read_text_catalog(catalog_csv, source_select_kwds={}, return_table=False):
+def read_gleam_catalog(
```

Comment:
I just realized that this method needs to be added to the __all__ list at the beginning of the file for it to be included in the module namespace.

It would be much cleaner to modify SkyModel to allow empty initialization and then make the read functions methods on that object (like how UVData operates). We should think about doing that in the future.

Reply:
We should talk about. I've been thinking it would be nice to be able to initialize a UVData object with a file.